src/main/scala/mesosphere/marathon/core/instance/Instance.scala

```diff
@@ -246,23 +274,39 @@ object Instance {
     // instanceId = $runSpecId.(instance-|marathon-)$uuid
     val InstanceIdRegex: Regex = """^(.+)\.(instance-|marathon-)([^\.]+)$""".r
 
+    private val ReservationIdRegex = """^(.+)([\._])([^_\.]+)$""".r
+
     private val uuidGenerator = Generators.timeBasedGenerator(EthernetAddress.fromInterface())
 
-    def runSpecId(instanceId: String): PathId = {
-      instanceId match {
-        case InstanceIdRegex(runSpecId, prefix, uuid) => PathId.fromSafePath(runSpecId)
-        case _ => throw new MatchError("unable to extract runSpecId from instanceId " + instanceId)
+    def forRunSpec(id: PathId): Id = Instance.Id(id, PrefixInstance, uuidGenerator.generate())
+
+    def fromIdString(idString: String): Instance.Id = {
+      idString match {
+        case InstanceIdRegex(safeRunSpecId, prefix, uuid) =>
+          val runSpec = PathId.fromSafePath(safeRunSpecId)
+          Id(runSpec, Prefix.fromString(prefix), UUID.fromString(uuid))
+        case _ => throw new MatchError(s"instance id $idString is not a valid identifier")
       }
     }
 
-    private def executorIdString(instanceId: String): String = {
-      instanceId match {
-        case InstanceIdRegex(runSpecId, prefix, uuid) => prefix + runSpecId + "." + uuid
-        case _ => throw new MatchError("unable to extract executorId from instanceId " + instanceId)
+    /**
+      * Extract instance id from reservation ids which were generated by [[Task.Id.reservationId]]
+      *
+      * @param reservationId The raw reservation id.
+      * @return The instance identifier belonging to the reservation.
+      */
+    def fromReservationId(reservationId: String): Instance.Id = {
+      reservationId match {
+        case InstanceIdRegex(safeRunSpecId, prefix, uuid) =>
+          val runSpec = PathId.fromSafePath(safeRunSpecId)
+          Id(runSpec, Prefix.fromString(prefix), UUID.fromString(uuid))
+        case ReservationIdRegex(safeRunSpecId, separator, uuid) =>
+          val runSpec = PathId.fromSafePath(safeRunSpecId)
+          Id(runSpec, PrefixMarathon, UUID.fromString(uuid))
```

Comment:
I am not sure if it is an issue, but in this case `uuid` is a task ID, therefore we need to make sure that it doesn't introduce a regression here. Essentially, an ID is always constructed for a non-existing instance.

Reply:
In this diff the reservations ids are always the instance ids. I'll have to check manually though.